### PR TITLE
feat(massmailing): #MA-725 duplicate massmailing when there is many p…

### PR DESCRIPTION
**Pull request by qpic-cgi**
## Describe your changes
Addition of multiple mailings by students and by relatives when there is more than one punishment or sanction.

## Checklist tests
On massmailing view, click on sending mail (or sms, or PDF).
- check multiple sending option:
  - try to send for a student that have multiple punishments (and/or) sanctions
  - try to send for a student that have only one  punishment (and/or) sanction
- uncheck multiple sending option:
  - try to send for a student that have multiple punishments (and/or) sanctions

## Issue ticket number and link
[Jira - MA-725](https://entsupport.gdapublic.fr/browse/MA-725)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [x] If it is a consequent feature, I have added thorough tests.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: massmailing/src/main/java/fr/openent/massmailing/mailing/MassMailingProcessor.java [hunk 1]
```diff
@@ -253,34 +250,15 @@ private JsonArray mergeDetentionWithSlots(JsonArray punishments) {
...
                                     .collect(Collectors.toList())
                     );
-
-                    item.put(Field.PUNISHMENTS, mergedPunishments);
+                    return mergedPunishments.stream();
                 })
-                .collect(Collectors.toList()));
-    }
-
-    private void formatPunishmentEvents(JsonObject massmailingHourEvents) {
```
Comment (qpic-cgi) [inline massmailing/src/main/java/fr/openent/massmailing/mailing/MassMailingProcessor.java:269]: le mapping des Punitions / Sanctions se fait maintenant directement avec la méthode duplicateMailingFromPunishments

File: massmailing/src/main/java/fr/openent/massmailing/mailing/MassMailingProcessor.java [hunk 2]
```diff
@@ -577,29 +609,6 @@ private List<String> getStudentsList() {
         return new ArrayList<>(students.getMap().keySet());
     }
 
-    /**
-     * Transform mailingTypeList object to an ArrayList containing each EventType code
-     *
-     * @return Event Type code list
-     */
-    private List<Integer> getEventTypeCodeList() {
```
Comment (qpic-cgi) [inline massmailing/src/main/java/fr/openent/massmailing/mailing/MassMailingProcessor.java:585]: plus utilisé

File: massmailing/src/test/java/fr/openent/massmailing/mailing/MassMailingProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        JsonObject massmail = new JsonObject()
+                .put("receiver_data", "test")
+                .put(Field.EVENTS,
+                        new JsonObject().put(MassmailingType.PUNISHMENT.name(), new JsonArray().add(
+                                new JsonObject().put(Field.PUNISHMENTS, new JsonArray()
+                                        .add(simplePunishment())
+                                        .add(groupedPunishment1())
+                                        .add(groupedPunishment2())
+                                )))
+                );
```
Comment (qpic-cgi) [inline massmailing/src/test/java/fr/openent/massmailing/mailing/MassMailingProcessorTest.java:47]: ici je ne peux pas utiliser la valeur stringifié a cause d'un problème de casting (LinkedHashmap)

**Review by JinDouang — APPROVED** [review]

File: massmailing/src/main/resources/public/ts/controllers/home.ts [hunk 4]
```diff
@@ -392,6 +394,16 @@ export const homeController = ng.controller('HomeController', ['$scope', 'route'
             }
         };
 
+        vm.getMassmailPunishmentsQuantity = (): number => {
+            return vm.massmailing.students
+                .filter((student:  MassmailingStudent) => student.selected)
```
Comment (JinDouang) [inline massmailing/src/main/resources/public/ts/controllers/home.ts]: espace en trop

File: massmailing/src/main/resources/public/ts/controllers/home.ts [hunk 5]
```diff
@@ -392,6 +394,16 @@ export const homeController = ng.controller('HomeController', ['$scope', 'route'
...
+                .map((student:  MassmailingStudent) =>
```
Comment (JinDouang) [inline massmailing/src/main/resources/public/ts/controllers/home.ts]: espace en trop

Comment (qpic-cgi) [reply massmailing/src/main/resources/public/ts/controllers/home.ts]: nhaaaaaa abusé X)

File: massmailing/src/main/resources/public/ts/controllers/home.ts [hunk 6]
```diff
@@ -302,14 +304,14 @@ export const homeController = ng.controller('HomeController', ['$scope', 'route'
             vm.formFilter.allAbsenceReasons = !vm.formFilter.allAbsenceReasons;
             vm.formFilter.noReasons = vm.formFilter.allAbsenceReasons;
 
-            vm.getAbsenceReasons().forEach((reason:Reason) => vm.formFilter.reasons[reason.id] = vm.formFilter.allAbsenceReasons)
+            vm.getAbsenceReasons().forEach((reason: Reason) => vm.formFilter.reasons[reason.id] = vm.formFilter.allAbsenceReasons)
```
Comment (qpic-cgi) [inline massmailing/src/main/resources/public/ts/controllers/home.ts]: indent

File: massmailing/src/main/resources/public/ts/controllers/home.ts [hunk 7]
```diff
@@ -302,14 +304,14 @@ export const homeController = ng.controller('HomeController', ['$scope', 'route'
...
         };
 
         vm.switchAllLatenessReasons = function (): void {
             vm.formFilter.allLatenessReasons = !vm.formFilter.allLatenessReasons;
             vm.formFilter.noLatenessReasons = vm.formFilter.allLatenessReasons;
 
-            vm.getLatenessReasons().forEach((reason:Reason) => vm.formFilter.reasons[reason.id] = vm.formFilter.allLatenessReasons)
+            vm.getLatenessReasons().forEach((reason: Reason) => vm.formFilter.reasons[reason.id] = vm.formFilter.allLatenessReasons)
```
Comment (qpic-cgi) [inline massmailing/src/main/resources/public/ts/controllers/home.ts]: indent

File: massmailing/src/main/resources/public/ts/controllers/home.ts [hunk 8]
```diff
@@ -561,7 +573,7 @@ export const homeController = ng.controller('HomeController', ['$scope', 'route'
                 return !bool;
             }
 
-            function filtredReasonIsFalse(map: { [key: number]: boolean }, reasons: Reason[] ): boolean {
+            function filtredReasonIsFalse(map: { [key: number]: boolean }, reasons: Reason[]): boolean {
```
Comment (qpic-cgi) [inline massmailing/src/main/resources/public/ts/controllers/home.ts]: indent

**Review by JinDouang — APPROVED** [review]

**Review by JinDouang — APPROVED** [review]

**Review by BastienBenard — APPROVED** [review]

**Review by flmartineau — APPROVED** [review]

**Review by JinDouang — APPROVED** [review]

**Pull request merged**